libdnf/repo/repo_sack.cpp

```diff
@@ -610,43 +612,80 @@ void RepoSack::fix_group_missing_xml() {
-                        xml_file_name.string());
-                    try {
-                        group.serialize(xml_file_name);
-                        xml_saved = true;
-                    } catch (utils::xml::XMLSaveError & ex) {
-                        logger.debug(ex.what());
-                    }
-                    if (xml_saved) {
-                        solv_repo->read_group_solvable_from_xml(xml_file_name);
+        if (!group_missing_xml.empty()) {
+            libdnf::comps::GroupQuery available_groups(base);
+            available_groups.filter_installed(false);
+            for (const auto & group_id : group_missing_xml) {
+                bool xml_saved = false;
+                if (directory_exists) {
+                    // try to find the repoid in availables
+                    libdnf::comps::GroupQuery group_query(available_groups);
+                    group_query.filter_groupid(group_id);
+                    if (group_query.size() == 1) {
+                        // GroupQuery is basically a set thus iterators and `.get()` method
+                        // return `const Group` objects.
+                        // To call non-const serialize method we need to make a copy here.
+                        libdnf::comps::Group group = group_query.get();
+                        auto xml_file_name = comps_xml_dir / (group_id + ".xml");
+                        logger.debug(
+                            "Re-creating installed group \"{}\" definition to file \"{}\".",
+                            group_id,
+                            xml_file_name.string());
+                        try {
+                            group.serialize(xml_file_name);
+                            xml_saved = true;
+                        } catch (utils::xml::XMLSaveError & ex) {
+                            logger.debug(ex.what());
+                        }
+                        if (xml_saved) {
+                            solv_repo->read_group_solvable_from_xml(xml_file_name);
+                        }
                     }
                 }
+                if (!xml_saved) {
+                    // fall-back to creating solvables only from system state
+                    solv_repo->create_group_solvable(group_id, system_state.get_group_state(group_id));
+                }
             }
-            if (!xml_saved) {
-                // fall-back to creating solvables only from system state
-                solv_repo->create_group_solvable(group_id, system_state.get_group_state(group_id));
+        }
+        if (!environments_missing_xml.empty()) {
+            libdnf::comps::EnvironmentQuery available_environments(base);
+            available_environments.filter_installed(false);
+            for (const auto & environment_id : environments_missing_xml) {
+                bool xml_saved = false;
+                if (directory_exists) {
+                    // try to find the repoid in availables
```

Comment:
Both comments for groups and environments probably shouldn't say `repoid` here?

Reply:
Yeah, sorry. Fixed.